tests/components/unifi/test_controller.py

```diff
@@ -290,33 +321,124 @@ async def test_reset_after_successful_setup(hass, aioclient_mock):
+        "ip": "10.0.0.1",
+        "is_wired": True,
+        "last_seen": dt_util.as_timestamp(dt_util.utcnow()),
+        "mac": "00:00:00:00:00:01",
+    }
+    await setup_unifi_integration(hass, aioclient_mock, clients_response=[client])
+
+    # Controller is connected
+    assert hass.states.get("device_tracker.client").state == "home"
+
+    mock_unifi_websocket(state=STATE_DISCONNECTED)
+    await hass.async_block_till_done()
+
+    # Controller is disconnected
+    assert hass.states.get("device_tracker.client").state == "unavailable"
+
+    mock_unifi_websocket(state=STATE_RUNNING)
+    await hass.async_block_till_done()
+
+    # Controller is once again connected
+    assert hass.states.get("device_tracker.client").state == "home"
+
+
 async def test_wireless_client_event_calls_update_wireless_devices(
-    hass, aioclient_mock
+    hass, aioclient_mock, mock_unifi_websocket
 ):
     """Call update_wireless_devices method when receiving wireless client event."""
-    config_entry = await setup_unifi_integration(hass, aioclient_mock)
-    controller = hass.data[UNIFI_DOMAIN][config_entry.entry_id]
+    await setup_unifi_integration(hass, aioclient_mock)
 
     with patch(
         "homeassistant.components.unifi.controller.UniFiController.update_wireless_clients",
         return_value=None,
     ) as wireless_clients_mock:
-        controller.api.websocket._data = {
-            "meta": {"rc": "ok", "message": "events"},
-            "data": [
-                {
-                    "datetime": "2020-01-20T19:37:04Z",
-                    "key": aiounifi.events.WIRELESS_CLIENT_CONNECTED,
-                    "msg": "User[11:22:33:44:55:66] has connected to WLAN",
-                    "time": 1579549024893,
-                }
-            ],
-        }
-        controller.api.session_handler("data")
+        mock_unifi_websocket(
+            data={
+                "meta": {"rc": "ok", "message": "events"},
+                "data": [
+                    {
+                        "datetime": "2020-01-20T19:37:04Z",
+                        "key": aiounifi.events.WIRELESS_CLIENT_CONNECTED,
+                        "msg": "User[11:22:33:44:55:66] has connected to WLAN",
+                        "time": 1579549024893,
+                    }
+                ],
+            },
+        )
 
         assert wireless_clients_mock.assert_called_once
 
 
+async def test_reconnect_mechanism(hass, aioclient_mock, mock_unifi_websocket, caplog):
+    """Verify reconnect prints only on first reconnection try."""
+    await setup_unifi_integration(hass, aioclient_mock)
+
+    caplog.clear()
+    aioclient_mock.post(f"https://{DEFAULT_HOST}:1234/api/login", status=502)
+
+    mock_unifi_websocket(state=STATE_DISCONNECTED)
+    await hass.async_block_till_done()
+
+    assert len(caplog.records) == 1
+
+    new_time = dt_util.utcnow() + timedelta(seconds=RETRY_TIMER)
+    async_fire_time_changed(hass, new_time)
+
+    assert len(caplog.records) == 2
```

Comment:
Can we assert something besides the logs, eg requests being made to reconnect?

Reply:
It defeats the purpose of verifying that it only logs on first reconnect attempt but sure